### PR TITLE
KAFKA-13498: keeping track of latest seen position in different state stores

### PR DESCRIPTION
**Pull request by patrickstuedi**
Previously [KAFKA-13480](https://github.com/apache/kafka/pull/11514) was adding position tracking to the RocksDB and InMemory KV store. This PR adds other stores such as CachingKVStore, and all Session and Windowed variants.

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/state/internals/InMemorySessionStoreTest.java [hunk 1]
```diff
@@ -83,4 +88,31 @@ public void shouldNotExpireFromOpenIterator() {
...
+
+        final Position expected = Position.emptyPosition();
+        long offset = 0;
+        for (final KeyValue<Windowed<String>, Long> k : entries) {
+            sessionStore.put(k.key, k.value);
+            expected.update("input", 0, offset);
+            offset++;
+        }
+
+        final MeteredSessionStore<String, Long> meteredSessionStore = (MeteredSessionStore<String, Long>) sessionStore;
```
Comment (patrickstuedi) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/InMemorySessionStoreTest.java:110]: This is not nice. One way to fix this would be to add getPosition() to the StateStore interface.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java [hunk 2]
```diff
@@ -131,6 +149,11 @@ public void put(final Bytes key, final byte[] value, final long windowStartTimes
                 });
             }
         }
+
+        if (stateStoreContext != null && stateStoreContext.recordMetadata().isPresent()) {
```
Comment (patrickstuedi) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java:153]: Optimistically updating position. Position makes sure we never move back. Or should we move this into the "value != null" case?

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java:153]: I think we'll still want to advance the position, even if we otherwise ignore the record.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Position.java [hunk 3]
```diff
@@ -44,10 +44,12 @@ private Position(final Map<String, Map<Integer, Long>> other) {
...
     public Position update(final String topic, final int partition, final long offset) {
-        position.computeIfAbsent(topic, k -> new ConcurrentHashMap<>());
-        final ConcurrentMap<Integer, AtomicLong> topicMap = position.get(topic);
-        topicMap.computeIfAbsent(partition, k -> new AtomicLong(0));
-        topicMap.get(partition).getAndAccumulate(offset, Math::max);
+        if (topic != null) {
+            position.computeIfAbsent(topic, k -> new ConcurrentHashMap<>());
+            final ConcurrentMap<Integer, AtomicLong> topicMap = position.get(topic);
+            topicMap.computeIfAbsent(partition, k -> new AtomicLong(0));
+            topicMap.get(partition).getAndAccumulate(offset, Math::max);
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Position.java:51]: Computing the max is something I didn't add in the framework PR. I left it up to the caller, but it looks like pretty much every caller would have to ensure it all over the place, so I'll go ahead and update https://github.com/apache/kafka/pull/11557 to preserve these semantics.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/InMemorySessionStoreTest.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/InMemorySessionStoreTest.java:110]: Thanks. I had this on my mind when I was working on the tests for the framework PR. It occurred to me that another approach would be to simply run a dummy query and check the returned position.

But I guess that doesn't really count as a "unit" test, since the intervening store layers could in theory be changing the position as well, so that approach might not be testing what it thinks it's testing.

This is more direct, and to be honest, I don't think it's that bad in the context of a unit test.

**Review by vvcephei — APPROVED** [review]
LGTM! Thanks, @patrickstuedi

**Comment by vvcephei** [comment]
I think the test build failed before because of that bug on trunk. I kicked it off again.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/InMemorySessionStoreTest.java (hunk 1 above)
Comment (vpapavas) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/InMemorySessionStoreTest.java:110]: I agree,  I don't think it is bad in the context of a unit test

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractSessionBytesStoreTest.java [hunk 4]
```diff
@@ -77,7 +77,7 @@
 
     private MockRecordCollector recordCollector;
 
-    private InternalMockProcessorContext context;
```
Comment (vpapavas) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractSessionBytesStoreTest.java:80]: Why was this needed?

**Review by vpapavas — APPROVED** [review]
Thank you @patrickstuedi ! LGTM

**Comment by vvcephei** [comment]
The test failures were known-flaky, unrelated tests.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractSessionBytesStoreTest.java (hunk 4 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/AbstractSessionBytesStoreTest.java:80]: It looks like @patrickstuedi needed to reference the context in the concrete test classes that subclass this.
